Quickstart doc for YugabyteDB PGX Go driver (#11905)
yugabyte/yugabyte-db: docs/content/latest/quick-start/build-apps/go/ysql-yb-pgx.md

    </a>
  </li>
  <li>
    <a href="../ycql/" class="nav-link">
      <i class="icon-cassandra" aria-hidden="true"></i>
      YCQL
    </a>
  </li>
</ul>

This driver is a fork of [jackc/pgx](https://github.com/jackc/pgx) driver, with additional feature of connection load balancing.

The following tutorial creates a simple Go application that connects to a YugabyteDB cluster using the [YugabyteDB pgx driver](https://pkg.go.dev/github.com/yugabyte/pgx), performs a few basic database operations — creating a table, inserting data, and running a SQL query — and then prints the results to the screen.

The first example specifically demonstrates the connection load balancing using the `pgx.Connect()` API whereas the second example demonstrates the same using `pgxpool.Connect()` API.

## Prerequisites

This tutorial assumes that:

- YugabyteDB is up and running. If you are new to YugabyteDB, you can download, install, and have YugabyteDB up and running within minutes by following the steps in [Quick start](../../../../quick-start/). Alternatively, you can use [Yugabyte Cloud](http://cloud.yugabyte.com/register) to get a fully managed database-as-a-service (DBaaS) for YugabyteDB.

> - [Go version 1.15](https://golang.org/dl/), or later, is installed.

Reviewer: ```suggestion
- [Go version 1.15](https://golang.org/dl/) or later is installed.
```

### SSL/TLS configuration

> You can find the details about the SSL/TLS configuration on the [YSQL PGX](../ysql-pgx/#ssl-tls-configuration) Go driver section. The same configuration is applicable to the YugabyteDB pgx driver as well.

Reviewer: ```suggestion
The YugabyteDB PGX driver SSL/TLS configuration is the same as for the [YSQL PGX](../ysql-pgx/#ssl-tls-configuration) Go driver.
```

### Go YugabyteDB driver

> The [Go YugabyteDB driver package (`pgx`)](https://pkg.go.dev/github.com/yugabyte/pgx) is a fork of jackc/pgx Go driver. The current release of `pgx v4` requires Go modules.

Reviewer: ```suggestion
The current release of the [Go YugabyteDB driver package (`pgx`)](https://pkg.go.dev/github.com/yugabyte/pgx) is v4, and requires Go modules.
```

To install the package locally, run the following commands:

```sh
mkdir yb-pgx
cd yb-pgx
go mod init hello
go get github.com/yugabyte/pgx/v4
go get github.com/yugabyte/pgx/v4/pgxpool
```

> ## Write an application with pgx.Connect()

Reviewer: Naming the anchor for easier linking :) 
```suggestion
## Write an application with pgx.Connect() {#pgx-connect-sample}
```

> Create a file `ybsql_hello_world.go` and copy the contents below.

Reviewer: ```suggestion
Create a file called `ybsql_hello_world.go` and paste the following contents into it:
```

```go
package main

import (

> 	"bufio"

Reviewer: Can you use spaces instead of hard tabs? It's likely to render more reliably, unless Go _requires_ tabs instead of spaces.

	fmt.Println("You can verify the connection counts on http://127.0.0.1:13000/rpcz and similar urls for other servers.")
	pause()
	closeConns(numconns)

	fmt.Println("---- Demonstrating topology-aware load balancing ----")
	url = fmt.Sprintf("%s?load_balance=true&topology_keys=cloud1.datacenter1.rack1", baseUrl)
	fmt.Printf("Connection url: %s\n", url)
	executeQueries(url)
	pause()
	closeConns(numconns)

	fmt.Println("Closing the application ...")
}

func closeConns(num int) {
	fmt.Printf("Closing %d connections ...\n", num)
	for i := 0; i < num; i++ {
		connCloseChan <- i
	}
}

func pause() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("\nPress Enter/return to proceed: ")
	reader.ReadString('\n')
}

func createTable(url string) {
	conn, err := pgx.Connect(context.Background(), url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close(context.Background())

	var dropStmt = `DROP TABLE IF EXISTS employee`
	_, err = conn.Exec(context.Background(), dropStmt)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Exec for drop table failed: %v\n", err)
	}

	var createStmt = `CREATE TABLE employee (id int PRIMARY KEY,

>                                              name varchar,

Reviewer: See what I mean about the rendering? This is what'll happen in the published docs too :)

		os.Exit(1)
	}

	// Read from the table.
	var name, language string
	var age int
	rows, err := conn.Query(context.Background(), "SELECT name, age, language FROM employee WHERE id = 1")
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()
	fstr := fmt.Sprintf("[%s] Query for id=1 returned: ", grid)
	for rows.Next() {
		err := rows.Scan(&name, &age, &language)
		if err != nil {
			log.Fatal(err)
		}
		fstr = fstr + fmt.Sprintf(" Row[%s, %d, %s]", name, age, language)
	}
	err = rows.Err()
	if err != nil {
		log.Fatal(err)
	}
	// log.Println(fstr)
	_, ok := <-ccChan
	if ok {
		conn.Close(context.Background())
	}
}

func printHostLoad() {
	for k, cli := range pgx.GetHostLoad() {
		str := "Current load on cluster (" + k + "): "
		for h, c := range cli {
			str = str + fmt.Sprintf("\n%-30s:%5d", h, c)
		}
		fmt.Println(str)
	}
}

func printAZInfo() {
	for k, zl := range pgx.GetAZInfo() {
		str := "Placement info details of cluster (" + k + "): "
		for z, hosts := range zl {
			str = str + fmt.Sprintf("\n    AZ [%s]: ", z)
			for _, s := range hosts {
				str = str + fmt.Sprintf("%s, ", s)
			}
		}
		fmt.Println(str)
	}
}
```

> The **const** values are set to the defaults for a local installation of YugabyteDB. If you are using Yugabyte Cloud, replace the **const** values in the file as follows:

Reviewer: ```suggestion
The **const** values are set to the defaults for a local installation of YugabyteDB. If you're using Yugabyte Cloud, replace the values as follows:
```

- **host** - The host address of your cluster. The host address is displayed on the cluster Settings tab.
- **user** - Your Yugabyte database username. In Yugabyte Cloud, the default user is **admin**.
- **password** - Your Yugabyte database password.
- **dbname** - The name of the Yugabyte database. The default Yugabyte database name is **yugabyte**.

**port** is set to 5433, which is the default port for the YSQL API.

### Enable SSL/TLS

For a Yugabyte Cloud cluster or a YugabyteDB cluster with SSL/TLS enabled, set the SSL-related environment variables as below.

```sh
$ export PGSSLMODE=verify-ca

> $ export PGSSLROOTCERT=~/root.crt  # Here, the CA certificate file is downloaded as `root.crt` under home directory. Modify your path accordingly.

Reviewer: ```suggestion
$ export PGSSLROOTCERT=~/root.crt
  # In this example, the CA certificate file is downloaded as `root.crt`
  # in your home directory. Modify your path accordingly.
```

```

### Run the application

```sh
$ go run ybsql_hello_world.go
```

This program expects user input to proceed through the application steps.

> For a local cluster with three servers, all of them with the placement info `cloud1.datacenter1.rack1`, you would see an output given below.

Reviewer: ```suggestion
For a local cluster with three servers, all of them with the placement info `cloud1.datacenter1.rack1`, you would see the following output:
```

```output
Connection url: postgres://yugabyte:yugabyte@localhost:5433/yugabyte?load_balance=true
Created table employee
Placement info details of cluster (127.0.0.1): 
    AZ [cloud1.datacenter1.rack1]: 127.0.0.3, 127.0.0.2, 127.0.0.1, 

Press Enter/return to proceed: 
---- Demonstrating uniform (cluster-aware) load balancing ----
Creating 12 connections ...
Current load on cluster (127.0.0.1): 
127.0.0.3                     :    4
127.0.0.2                     :    4
127.0.0.1                     :    4
You can verify the connection counts on http://127.0.0.1:13000/rpcz and similar urls for other servers.

Press Enter/return to proceed: 
Closing 12 connections ...
---- Demonstrating topology-aware load balancing ----
Connection url: postgres://yugabyte:yugabyte@localhost:5433/yugabyte?load_balance=true&topology_keys=cloud1.datacenter1.rack1
Creating 12 connections ...
Current load on cluster (127.0.0.1): 
127.0.0.1                     :    4
127.0.0.3                     :    4
127.0.0.2                     :    4

Press Enter/return to proceed: 
Closing 12 connections ...
Closing the application ...
```

> 

Reviewer: Kill extra blank lines... (1 is good, 2+ is not)
```suggestion
```

> ## Write an application with pgxpool.Connect()

Reviewer: ```suggestion
## Write an application with pgxpool.Connect() {#pgxpool-connect-sample}
```

> Create a file `ybsql_hello_world_pool.go` and copy the contents below.

Reviewer: ```suggestion
Create a file called `ybsql_hello_world_pool.go` and paste the following contents into it:
```

	for {
		// Read from the table.
		var name, language string
		var age int
		rows, err := pool.Query(context.Background(), "SELECT name, age, language FROM employee WHERE id = 1")
		if err != nil {
			log.Fatalf("pool.Query() failed, %s", err)
		}
		defer rows.Close()
		fstr := fmt.Sprintf("[%s] Query for id=1 returned: ", grid)
		for rows.Next() {
			err := rows.Scan(&name, &age, &language)
			if err != nil {
				log.Fatalf("rows.Scan() failed, %s", err)
			}
			fstr = fstr + fmt.Sprintf(" Row[%s, %d, %s] ", name, age, language)
		}
		err = rows.Err()
		if err != nil {
			fmt.Printf("%s, retrying ...\n", err)
			continue
		}
		time.Sleep(5 * time.Second)
		break
	}
}

func printHostLoad() {
	for k, cli := range pgx.GetHostLoad() {
		str := "Current load on cluster (" + k + "): "
		for h, c := range cli {
			str = str + fmt.Sprintf("\n%-30s:%5d", h, c)
		}
		fmt.Println(str)
	}
}

func printAZInfo() {
	for k, zl := range pgx.GetAZInfo() {
		str := "Placement info details of cluster (" + k + "): "
		for z, hosts := range zl {
			str = str + fmt.Sprintf("\n    AZ [%s]: ", z)
			for _, s := range hosts {
				str = str + fmt.Sprintf("%s, ", s)
			}
		}
		fmt.Println(str)
	}
}
```

> The **const** values are set to the defaults for a local installation of YugabyteDB. If you are using Yugabyte Cloud, replace the **const** values in the file as explained earlier.

Reviewer: ```suggestion
The **const** values are set to the defaults for a local installation of YugabyteDB. If you are using Yugabyte Cloud, replace the **const** values in the file as [explained earlier](#pgx-connect-sample).
```

### Enable SSL/TLS

> For a Yugabyte Cloud cluster or a YugabyteDB cluster with SSL/TLS enabled, set the SSL-related environment variables as below.

Reviewer: ```suggestion
For a Yugabyte Cloud cluster or a YugabyteDB cluster with SSL/TLS enabled, set the SSL-related environment variables as follows:
```

```sh
$ export PGSSLMODE=verify-ca

> $ export PGSSLROOTCERT=~/root.crt  # Here, the CA certificate file is downloaded as `root.crt` under home directory. Modify your path accordingly.

Reviewer: ```suggestion
$ export PGSSLROOTCERT=~/root.crt
  # In this example, the CA certificate file is downloaded as `root.crt`
  # in your home directory. Modify your path accordingly.
```

```

### Run the application

```sh
$ go run ybsql_hello_world_pool.go
```

This program expects user input to proceed through the application steps.

> For a local cluster with three servers, all of them with the placement info `cloud1.datacenter1.rack1`, you would see an output given below.

Reviewer: ```suggestion
For a local cluster with three servers, all of them with the placement info `cloud1.datacenter1.rack1`, you would see the following output:
```
